File: problem_081.c
// -*- compile-command: "gcc -o problem_081 problem_081.c -Wall" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

FILE *matrixFile;

int main(){
    printf("Project Euler - Problem 81:\n"
           "Find the minimal path sum in a 31K text file containing a 80 by 80 matrix, from the top left to the bottom right by only moving right and down.\n\n");

    int matrix[80][80]={};
    int x = 0, y = 0;

    //Populate table from text file
    matrixFile = fopen ("problem_081.txt", "rt");

    char ch;
    while((ch = fgetc(matrixFile)))
    {
        if (ch == EOF) break;
        if (ch == ',')
        {
            x++;
            continue;
        }

        if (ch == '\n')
        {
            x = 0;
            y++;
            continue;
        }

        matrix[y][x] *= 10;
        matrix[y][x] += ch - '0';
    }

    fclose(matrixFile);

    //Go through entire grid
    for (x=0; x<80; x++)
    {
        for (y=0; y<80; y++)
	{
            //Find best move for each position
            if (x==0 && y==0) continue; //Skip first cell

            //If on left or top edge, make obligatory move
            if (x==0)
                matrix[x][y] += matrix[x][y-1];
            else if (y==0)
                matrix[x][y] += matrix[x-1][y];

            //Find best move, and make it, replace element value
            else if (matrix[x-1][y] < matrix[x][y-1])
                matrix[x][y] += matrix[x-1][y];
            else
                matrix[x][y] += matrix[x][y-1];
	}
    }

    printf("Smallest path sum: %d\n", matrix[79][79]);
    return 0;
}

File: problem_081.rs
// -*- compile-command: "rustc -o problem_081_rs problem_081.rs" -*-
// Copyright (c) 2016 Michael Caldwell
use std::fs::File;
use std::io::Read;

fn main() {
    println!("Project Euler - Problem 81:");
    println!("Find the minimal path sum in a 31K text file containing a 80 by 80 matrix, from the top left to the bottom right by only moving right and down.\n");

    let mut matrix: [[u32; 80]; 80] = [[0; 80]; 80];

    // Import data from file
    let mut raw_data = String::new();
    let mut f = File::open("problem_081.txt").unwrap();
    f.read_to_string(&mut raw_data).unwrap();

    // Split data lines into vector
    let data: Vec<&str> = raw_data.split_whitespace().collect();

    // Put vector into matrix array
    for row in 0..data.len() {
        // Split line into vector of values
        let line_of_data: Vec<&str> = data[row].split(",").collect();

        for col in 0..line_of_data.len() {
            matrix[row][col] = line_of_data[col].parse::<u32>().unwrap();
        }
    }

    // Solve!
    for y in 0..matrix.len() {
        for x in 0..matrix.len() {
            // Case 0: Top-left corner
            if y == 0 && x == 0 { continue; };

            // Case 1: Along top edge
            if y == 0 {
                matrix[y][x] += matrix[y][x-1];
                continue;
            }

            // Case 2: Along left edge
            if x == 0 {
                matrix[y][x] += matrix[y-1][x];
                continue;
            }

            // Case 3: Can choose
            if matrix[y-1][x] < matrix[y][x-1] {
                matrix[y][x] += matrix[y-1][x];
            } else {
                matrix[y][x] += matrix[y][x-1];
            }
        }
    }

    // Print last cell for answer
    println!("{}", matrix[matrix.len()-1][matrix.len()-1]);
}
